Replace float16 with at::Half in caffe2

li-roy opened the pull request:
Summary: The first commit uses at::Half as a replacement for caffe2::float16. float16 conversions and arithmetic removed to make use of Half equivalents. Also, as a side effect, aten_op works for gpu half now.

The second commit is a codemod (caffe2::float16 -> at::Half). Notable changes: `caffe2/perfkernels/embedding_lookup.cc`, the macro was changed because at::Half can't be in a function name. Same thing for the codegen in `hp_emblookup_codegen.py`.

Differential Revision: D9892158

ezyang commented:
Errr, this wasn't just a simple codemod, was it? Can we have a billing of changes?

ezyang commented on caffe2/perfkernels/embedding_lookup.cc:
```diff
@@ -82,92 +82,92 @@ static void EmbeddingLookupGenericSlow(
-        normalize_by_lengths,                                                              \
-        out);                                                                              \
+#define EMBEDDING_SPECIALIZATION(                                                                      \
+    IndexTypeName, IndexType, InTypeName, InType, OutTypeName, OutType, IS_WEIGHT_POSITIONAL)          \
```
Hm?

li-roy replied:
InType was used in function names, but we can't use at::Half there, so I made InTypeName to use in the function name and kept InType for the actual type. We don't really need IndexTypeName or OutTypeName because those are never Half, but I thought it would be strange not to have them.

e.g. ```EmbeddingLookup_##IndexTypeName##_##InTypeName##_##OutTypeName##_##IS_WEIGHT_POSITIONAL##__base```

Alternatively, we could just `using at::Half` in this file, but I kind of want it match `hp_emblookup_codegen.py`

ezyang commented on caffe2/perfkernels/embedding_lookup.cc:
```diff
@@ -82,92 +82,92 @@ static void EmbeddingLookupGenericSlow(
-EMBEDDING_SPECIALIZATION(int64_t, uint8_t, float, false);
+EMBEDDING_SPECIALIZATION(int32_t, int32_t, float, float, float, float, false);
+EMBEDDING_SPECIALIZATION(int64_t, int64_t, float, float, float, float, false);
+EMBEDDING_SPECIALIZATION(int32_t, int32_t, half, at::Half, float, float, false);
```
What's going on here?

li-roy replied:
Same as above.

ezyang commented on caffe2/python/examples/resnet50_trainer.py:
```diff
@@ -641,7 +641,7 @@ def main():
                         choices=['float', 'float16'],
                         help='Data type used for training')
-    parser.add_argument('--float16_compute', action='store_true',
+    parser.add_argument('--half_compute', action='store_true',
```
Euh... BC?

li-roy replied:
Yeah these python changes aren't supposed to be here. Thanks for catching it.

ezyang commented on caffe2/python/helpers/conv.py:
```diff
@@ -27,7 +27,7 @@ def _ConvBase(
     cudnn_exhaustive_search=False,
     ws_nbytes_limit=None,
-    float16_compute=False,
+    half_compute=False,
```
BC?

ezyang commented on caffe2/python/helpers/fc.py:
```diff
@@ -13,7 +13,7 @@
     model, op_call, blob_in, blob_out, dim_in, dim_out, weight_init=None,
         bias_init=None, WeightInitializer=None, BiasInitializer=None,
-        enable_tensor_core=False, float16_compute=False, **kwargs
+        enable_tensor_core=False, half_compute=False, **kwargs
```
BC?

ezyang commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -70,16 +69,16 @@ __global__ void SparseAdadeltaKernel(
+    float mom_delta_new = decay * param_mom_delta[paramIdx] +
         (1.0f - decay) * grad_new * grad_new;
-    mixed_store(&mom_delta_new, &(param_mom_delta[paramIdx]));
+    param_mom_delta[paramIdx] = mom_delta_new;
```
Yay!

ezyang commented on caffe2/utils/conversions.h:
```diff
@@ -21,218 +21,16 @@ namespace caffe2 {
 namespace convert {
 
-namespace {
-inline float16 cpu_float2half_rn(float f) {
```
If you want to be really careful about this, cross-test this with the ATen implementation to check if either have bugs :>

ezyang reviewed (commented):
I'm mostly a little concerned about Python side BC-breaking changes.

li-roy commented:
Yeah, I'll put out a list of changes tomorrow, just wanted to start running tests tonight.

gchanan commented on aten/src/ATen/core/typeid.h:
```diff
@@ -30,6 +30,13 @@
+// Make at::Half a fundamental type.
+namespace std {
+template<>
+struct is_fundamental<at::Half> : std::true_type {
```
do you need this?  at::Half is not fundamental.

li-roy replied:
Yeah, caffe2 needs this, I moved it from caffe2/core/types.h

gchanan commented on caffe2/operators/fully_connected_op_gpu.cc:
```diff
@@ -67,41 +67,41 @@ bool RunFullyConnectedGradientOpOnCUDADevice(
-  } else if (op->Input(0).template IsType<float16>()) {
-    if (float16_compute) {
+  } else if (op->Input(0).template IsType<at::Half>()) {
+    if (float_compute) {
```
shouldn't this be called half_compute?

li-roy replied:
I think I need to change them all back to float16_compute for bc

gchanan replied:
that's fine.

gchanan commented on caffe2/python/examples/resnet50_trainer.py:
```diff
@@ -391,7 +391,7 @@ def create_resnext_model_ops(model, loss_scale):
         stepsz = int(30 * args.epoch_size / total_batch_size / num_shards)
 
-        if args.float16_compute:
+        if args.float_compute:
```
shouldn't this be called half_compute?

li-roy commented:
@pytorchbot retest this please.

Commit:
Replace float16 with at::Half in caffe2 (#11676)

Summary:
- Finishes unifying Half type in pytorch and caffe2
- As a side effect, aten_op works for fp16 now
Pull Request resolved: https://github.com/pytorch/pytorch/pull/11676

Differential Revision: D9829019

fbshipit-source-id: e5f800024478c2e68ef29f4c06b4f0002f81a3f7

facebook-github-bot committed:
codemod: caffe::float16 -> at::Half (#11785)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/11785

Replace each instead of float16 with Half.

Reviewed By: Yangqing

Differential Revision: D9892158

fbshipit-source-id: c1f8dd9233423786b4a129569d4df2d39babd58e
